jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py

```diff
@@ -148,267 +164,202 @@ def user_api(self):
+            # NOTE: this assumes that all image data is coerced to surface brightness units
+            layers = [lyr for lyr in msg.viewer.layers if lyr.layer.data.label == msg.data.label]
+            self._handle_attribute_display_unit(self.sb_unit_selected, layers=layers)
+
+    @observe('spectral_unit_selected', 'flux_unit_selected',
+             'angle_unit_selected', 'sb_unit_selected',
+             'time_unit_selected')
+    def _on_unit_selected(self, msg):
         """
-        Observes toggle between surface brightness or flux selection for
-        spectrum viewer to trigger translation.
+        When any user selection is made, update the relevant viewer(s) with the new unit,
+        and then emit a GlobalDisplayUnitChanged message to notify other plugins of the change.
         """
+        if not len(msg.get('new', '')):
+            # empty string, nothing to set yet
+            return
 
-        if msg.get('name') == 'spectral_y_type_selected':
-            self._translate(self.spectral_y_type_selected)
+        axis = msg.get('name').split('_')[0]
 
-    @observe('flux_unit_selected')
-    def _on_flux_unit_changed(self, msg):
+        if axis == 'spectral':
+            xunit = _valid_glue_display_unit(self.spectral_unit.selected, self.spectrum_viewer, 'x')
+            self.spectrum_viewer.state.x_display_unit = xunit
+            self.spectrum_viewer.set_plot_axes()
+
+        elif axis == 'flux':
+            if len(self.angle_unit_selected):
+                # NOTE: setting sb_unit_selected will call this method again with axis=='sb',
+                # which in turn will call _handle_attribute_display_unit,
+                # _handle_spectral_y_unit (if spectral_y_type_selected == 'Surface Brightness'),
+                #  and send a second GlobalDisplayUnitChanged message for sb
+                self.sb_unit_selected = _flux_to_sb_unit(self.flux_unit.selected,
+                                                         self.angle_unit.selected)
+
+            if self.spectral_y_type_selected == 'Flux':
+                self._handle_spectral_y_unit()
+
+        elif axis == 'angle':
+            if len(self.flux_unit_selected):
+                # NOTE: setting sb_unit_selected will call this method again with axis=='sb',
+                # which in turn will call _handle_attribute_display_unit,
+                # _handle_spectral_y_unit (if spectral_y_type_selected == 'Surface Brightness'),
+                #  and send a second GlobalDisplayUnitChanged message for sb
+                self.sb_unit_selected = _flux_to_sb_unit(self.flux_unit.selected,
+                                                         self.angle_unit.selected)
+
+        elif axis == 'sb':
+            self._handle_attribute_display_unit(self.sb_unit_selected)
+
+            if self.spectral_y_type_selected == 'Surface Brightness':
+                self._handle_spectral_y_unit()
+
+        elif axis == 'time':
+            pass
```

Comment:
Not needed?
```suggestion
```

Reply:
probably not, I can replace with a comment if you'd rather.  It's here just to remind us that time is not currently handled (even though it can exist).

Reply:
Comment would be less confusing unless you intend to throw warning or exception.